### PR TITLE
Improve `&&` and `&` syntax highlighting by altering `StringName` highlighting

### LINKED ISSUE #11381: Improve `&&` and `&` syntax highlighting by altering `StringName` highlighting
### Describe the project you are working on

A game, with the main text editor being Godot.

### Describe the problem or limitation you are having in your project

The `StringName`(and by proxy `&&(and)` and `&(bitwise and)`) syntax highlighting is a bit of a mess.
- On new lines(mainly using `\`), the first ampersand of `&&` or `&` are the color of `StringName` highlighting. This really irks me, it's the only reason I haven't used `&&` over `and`. As an anecdote, my first impression of seeing it in 4.x(videos online) is that godot was unprofessional and I couldn't understand why such a glaring flaw like that would hang around.
- It's misleading too. If the & is spaced away from the quotes, it stays the color of `StringName`, despite being parsed as `&` or else borked. Due to my unfamiliarity, that quirk made me think you could space out the & if you wanted to. If the syntax highlighting changed, it would be much harder to confuse.
- For some reason, the 3rd & in a row after `&&` is the color of `StringName` (not pictured). Just strange and I don't know why it's there other than just 'cause. It goes back to symbol color after the 3rd.

Image

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Simplifying `StringName` highlighting would be the best, just make the ampersand symbol colored unless directly next to `'`/`"`(and not inside of another color block like strings). If there are arguments to the contrary, I would hear them out.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Link to my PR: https://github.com/godotengine/godot/pull/100575
Here is the comparison. I think it's a more consistent coloration that most wouldn't notice, but advanced GDScript users should enjoy. As in, bitwise and using `&&` is on the rarer side, especially for newcomers.

Image

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, as far as I know this syntax highlighting is baked into the text editor. You could work around it by changing `StringName` color to match the symbol color, but now your `StringName`s look like symbols or vice versa. Not a good compromise.

### Is there a reason why this should be core and not an add-on in the asset library?

See above section.

### PR DESCRIPTION
**Pull request by Wierdox**
Closes https://github.com/godotengine/godot-proposals/issues/11381

Not much to say about this PR, new code should be fairly easy to understand. Check out the proposal for an explanation and some details. That said, it could probably be cleaned up a bit, I always seem to miss something.

Image

### CODE REVIEW

**Review by Mickeon — APPROVED** [review]
I approve in principle as this looks like an obvious improvement to what can be seen as an oversight. But someone savvy in GDScript should take a look at it, perhaps stress-test it.

File: modules/gdscript/editor/gdscript_highlighter.cpp [hunk 1]
```diff
@@ -560,12 +560,17 @@ Dictionary GDScriptSyntaxHighlighter::_get_line_syntax_highlighting_impl(int p_l
 			}
 		}
 
-		// Keep symbol color for binary '&&'. In the case of '&&&' use StringName color for the last ampersand.
+		// Set color of StringName, keeping symbol color for binary '&&' and '&'.
 		if (!in_string_name && in_region == -1 && str[j] == '&' && !is_binary_op) {
-			if (j >= 2 && str[j - 1] == '&' && str[j - 2] != '&' && prev_is_binary_op) {
-				is_binary_op = true;
-			} else if (j == 0 || (j > 0 && str[j - 1] != '&') || prev_is_binary_op) {
+			if (j + 1 <= line_length - 1 && (str[j + 1] == '\'' || str[j + 1] == '\"')) {
```
Comment (AThousandShips) [inline modules/gdscript/editor/gdscript_highlighter.cpp:565]:
```suggestion
			if (j + 1 <= line_length - 1 && (str[j + 1] == '\'' || str[j + 1] == '"')) {
```
Unnecessary escape

**Review by AThousandShips — APPROVED** [review]
The improvement is great

@dalexeev @HolonProduction do we have any testing facilities for the highlighter? If so there should probably be test cases for this, or if we are working on that functionality this should be accounted for if merged

**Comment by dalexeev** [comment]
> do we have any testing facilities for the highlighter?

* There is #88910, but I'm not sure how clear the test output is.

**Review by Repiteo — APPROVED** [review]
In the absence of a clear testing method, I'm fine approving this at face-value & covering potential regressions manually

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**